part/txn.go

```diff
@@ -401,75 +401,88 @@ func (txn *Txn[T]) delete(root *header[T], key []byte) (oldValue T, hadOld bool,
 		return
 	}
 
-	// The target was found, rebuild the tree from the root upwards.
+	// The target was found, rebuild the tree from the leaf upwards to the root.
 	parents[0].node = root
 
 	for i := len(parents) - 1; i > 0; i-- {
 		parent := &parents[i-1]
-		target := &parents[i]
-
-		// Clone the parent to mutate it.
-		parent.node = txn.cloneNode(parent.node)
-		children := parent.node.children()
-
-		if target.node == this && target.node.size() > 0 {
+		this := &parents[i]
+
+		switch {
+		case this.node != target:
+			// Update the parent to point to the new child cloned
+			// on the previous iteration.
+			parent.node = txn.cloneNode(parent.node)
+			parent.node.children()[this.index] = this.node
+		case this.node.size() > 0:
```

Comment:
suggestion: IIUC, this case and the `default` case are mutually exclusive and are hit exactly once in this loop - namely when `this` is `target`. Wouldn't it be clearer if you pulled out the handling of `target` into an `if target.node.size() > 0` before the loop (which then becomes simple)?